### PR TITLE
Refactors CRH and Commitment, improves performance

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR does not change the expected result of CRH or Commitment schemes. Rather, it removes inefficiencies that reduces the number of constraints.

- Improves Pedersen hashing by 4%, and BHP hashing by 41%
- Reduces the BHP CRH gadget constraint count by ~17%
- Fixes an incorrect enforcement of number of `bytes` as opposed to number of `bits`:
```
input.len() <= WINDOW_SIZE * NUM_WINDOWS // input is sometimes [u8], when we should be checking [bool]
```
- Fixes an incorrect enforcement of number of `bits`:
```
input.len() <= WINDOW_SIZE * NUM_WINDOWS // before
input.len() <= WINDOW_SIZE * NUM_WINDOWS * BHP_CHUNK_SIZE // after
```
- Fixes BHP input padding from a fixed constant to the actual input size.
    - Note: This did not change the native output, however it did reduce the number of constraints for the BHP hash circuit, thus requiring a resampling of parameters.
    
Remark: This PR improves transaction generation times by 4.7%.

### CODE REVIEW

File: algorithms/src/commitment/bhp.rs [hunk 1]
```diff
@@ -56,77 +51,25 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> Com
...
-        for (bit, power) in scalar_bits.take(G::ScalarField::size_in_bits()).zip_eq(&self.random_base) {
+        let scalar_bits = BitIteratorLE::new(randomness.to_repr()).take(G::ScalarField::size_in_bits());
+        for (bit, power) in scalar_bits.zip_eq(&self.random_base) {
             if bit {
                 output += power
             }
         }
 
-        let affine = output.into_affine();
-        debug_assert!(affine.is_in_correct_subgroup_assuming_on_curve());
```
Comment (Pratyush) [inline algorithms/src/commitment/bhp.rs]: Why'd we remove this debug assertion?

File: algorithms/src/crh/bhp.rs [hunk 2]
```diff
@@ -51,43 +40,69 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
...
-        }
-
-        let maximum_num_chunks_in_segment = calculate_num_chunks_in_segment::<G::ScalarField>();
-        if WINDOW_SIZE > maximum_num_chunks_in_segment {
-            panic!(
-                "BHP CRH must have a window size resulting in scalars < (p-1)/2, \
-                 maximum segment size is {}",
-                maximum_num_chunks_in_segment
-            );
+        debug_assert_eq!(BHP_CHUNK_SIZE, 3);
```
Comment (Pratyush) [inline algorithms/src/crh/bhp.rs]: we can move this to a compile time assertion:

```rust
const ASSERT: () = assert!(BHP_CHUNK_SIZE == 3);
```

File: algorithms/src/crh/bhp.rs [hunk 3]
```diff
@@ -96,164 +111,34 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
...
-
-    /// Precondition: number of elements in `input` == `num_bits`.
-    pub(crate) fn hash_bits_inner<S: Borrow<bool>>(
-        &self,
-        input: impl Iterator<Item = S>,
-        num_bits: usize,
-    ) -> Result<G, CRHError> {
-        if num_bits > WINDOW_SIZE * NUM_WINDOWS {
-            return Err(CRHError::IncorrectInputLength(num_bits, WINDOW_SIZE, NUM_WINDOWS));
+    pub(crate) fn hash_bits_inner(&self, input: &[bool]) -> Result<G, CRHError> {
```
Comment (Pratyush) [inline algorithms/src/crh/bhp.rs:112]: This should just output `G::BaseField`.

File: algorithms/src/crh/bhp.rs [hunk 4]
```diff
@@ -96,164 +111,34 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
...
+        // Ensure the input size is within the parameter size,
+        if input.len() > NUM_WINDOWS * WINDOW_SIZE * BHP_CHUNK_SIZE {
```
Comment (Pratyush) [inline algorithms/src/crh/bhp.rs:114]: This is different from the earlier input.len() check

File: algorithms/src/commitment/bhp.rs (hunk 1 above)
Comment (howardwu) [reply algorithms/src/commitment/bhp.rs]: I'll revert

Comment (howardwu) [reply algorithms/src/commitment/bhp.rs]: Done

File: algorithms/src/crh/bhp.rs [hunk 5]
```diff
@@ -96,164 +111,34 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
...
-        }
-        let base_lookup = self.base_lookup(&self.bases);
-        debug_assert_eq!(base_lookup.len(), NUM_WINDOWS);
-        for bases in base_lookup.iter() {
-            debug_assert_eq!(bases.len(), WINDOW_SIZE);
+        // Pad the input to a multiple of `BHP_CHUNK_SIZE` for hashing.
+        let mut input = input.to_vec();
+        if input.len() % BHP_CHUNK_SIZE != 0 {
+            let padding = BHP_CHUNK_SIZE - (input.len() % BHP_CHUNK_SIZE);
+            input.extend_from_slice(&vec![false; padding]);
```
Comment (Pratyush) [inline algorithms/src/crh/bhp.rs]:
```suggestion
            let padding = BHP_CHUNK_SIZE - (input.len() % BHP_CHUNK_SIZE);
            input.extend_from_slice(&[false; 3][..padding]);
```

Comment (Pratyush) [reply algorithms/src/crh/bhp.rs]: Avoids an allocation

File: algorithms/src/crh/bhp.rs [hunk 6]
```diff
@@ -96,164 +111,34 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
...
-        );
-        for bases in self.bases.iter() {
-            debug_assert_eq!(bases.len(), WINDOW_SIZE);
-        }
-        let base_lookup = self.base_lookup(&self.bases);
-        debug_assert_eq!(base_lookup.len(), NUM_WINDOWS);
-        for bases in base_lookup.iter() {
-            debug_assert_eq!(bases.len(), WINDOW_SIZE);
+        // Pad the input to a multiple of `BHP_CHUNK_SIZE` for hashing.
+        let mut input = input.to_vec();
```
Comment (Pratyush) [inline algorithms/src/crh/bhp.rs:119]: We might want to keep the stack allocation logic

File: algorithms/src/crh/bhp.rs (hunk 3 above)
Comment (howardwu) [reply algorithms/src/crh/bhp.rs:112]: The BHP commitment needs the output to compute the randomness on top of it.

File: algorithms/src/crh/bhp.rs (hunk 4 above)
Comment (howardwu) [reply algorithms/src/crh/bhp.rs:114]: Wdym?

File: algorithms/src/crh/bhp.rs (hunk 5 above)
Comment (howardwu) [reply algorithms/src/crh/bhp.rs]: Done

File: algorithms/src/crh/bhp.rs (hunk 6 above)
Comment (howardwu) [reply algorithms/src/crh/bhp.rs:119]: The stack allocation logic requires a worst case allocation, which ends up being too expensive.

File: algorithms/src/crh/bhp.rs (hunk 3 above)
Comment (Pratyush) [reply algorithms/src/crh/bhp.rs:112]: ah good point.

**Comment by howardwu** [comment]
I ended up removing `const ASSERT: () = assert!(BHP_CHUNK_SIZE == 3);` since it's now trivially true.

File: algorithms/src/crh/bhp.rs (hunk 4 above)
Comment (Pratyush) [reply algorithms/src/crh/bhp.rs:114]: Sorry, didn't realize you noted that already in the PR description. this aligns with the arkworks code.

**Review by Pratyush — APPROVED** [review]

**Pull request merged**